File: reader/csharp/DataAccess.cs
using ProtoBuf;
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using UnityEngine;

#if LUA_SUPPORT
using LuaInterface;
#endif

internal class T_LOG
{
    public static void Log(string info)
    {
        Debug.Log(info);
    }
    
    public static void LogError(string info)
    {
        Debug.LogError(info);
    }
}

public class PBDataModel
{
    private Dictionary<string,object> _converted = new Dictionary<string, object>();

    protected object GetConvertData(string fieldName, object value, string alias, bool cachable)
    {
        if(cachable && _converted.ContainsKey(fieldName))
        {
            return _converted[fieldName];
        }

        if(DataAccess.DataConvertHandler == null)
        {
            throw new Exception($"convert field {fieldName} value need a convetor");
        }

        var data = DataAccess.DataConvertHandler(this, fieldName, value, alias, cachable);
        if (cachable)
        {
            _converted[fieldName] = data;
        }
        return data;
    }     
    
    public object Clone()
    {
        var type = GetType();
        var clone = Activator.CreateInstance(type);
        var props = type.GetProperties();
        for(var i=0;i<props.Length;i++)
        {
            var prop = props[i];
            var attris = prop.GetCustomAttributes(false);
            for(var ai = 0; ai < attris.Length; ai++)
            {
                var attri = attris[ai];
                if(attri is ProtoMemberAttribute)
                {
                    prop.SetValue(clone, prop.GetValue(this));

                    break;
                }
            }
        }
        return clone;
    }

    #if LUA_SUPPORT
    public virtual LuaTable ToLuaTable()
    {
        if (DataAccess.luaState == null)
        {
            Debug.LogError("must set luastate first!");
        }
        
        var table = DataAccess.luaState.NewTable();
        var type = GetType();
        var props = type.GetProperties();
        for(var i=0;i<props.Length;i++)
        {
            var prop = props[i];
            var attris = prop.GetCustomAttributes(false);
            for(var ai = 0; ai < attris.Length; ai++)
            {
                var attri = attris[ai];
                if(attri is ProtoMemberAttribute)
                {
                    if(DataAccess.UseProtoMemberTagAsHashtableKey)
                    {
                        table[(attri as ProtoMemberAttribute).Tag] = prop.GetValue(this);
                    }
                    else
                    {
                        table[prop.Name] = prop.GetValue(this);
                    }

                    break;
                }
            }
        }
        return table;
    }
    #endif
    
    public object GetPropValue(string name)
    {
        var tp = GetType();
        return tp.GetProperty(name).GetValue(this);
    }
}

public class PBDataModels
{
    
}

public delegate string FileNameGenerateHandler(string typeName);
public delegate byte[] LoadDataHandler(string datafile);

public partial class DataAccess
{
    #if LUA_SUPPORT
    public static LuaState luaState;
    #endif
    
    /// <summary>
    /// 是否使用ProtoMember的tag作为hashtable的key
    /// </summary>
    public static bool UseProtoMemberTagAsHashtableKey = false;
    public static bool CacheHashValue = true;
    public static string DataExt = ".bytes";
    public const string DefaultIDName = "Id";

    /// <summary>
    /// 数据文件以assets为根目录的路径
    /// </summary>
    public static string DataDir { get; private set; }
    public static LoadDataHandler Loader { get; private set; }
    public static FileNameGenerateHandler Generator { get; private set; }

    public delegate object ConvertHandler(object item, string field, object data, string alias, bool cachable);
    public static ConvertHandler DataConvertHandler;

    public static void Initial(string dataDir, LoadDataHandler loadHandle = null, FileNameGenerateHandler fileNameGenerateHandle = null)
    {
        DataDir = dataDir;
        Loader = loadHandle;
        Generator = fileNameGenerateHandle;
    }
}

public partial class DataContainer<TItem> : DataContainer
    where TItem : PBDataModel
{
    protected LoadDataHandler localLoader;
    protected FileNameGenerateHandler localGenerator;

    private string OnGenerateFilename(string typeName)
    {
        if (DataAccess.DataDir == null)
        {
            Debug.LogError("you need initial data access first!");
            return typeName.ToLower();
        }
        
        if (localGenerator != null)
        {
            return localGenerator(typeName);
        }

        if (DataAccess.Generator != null)
        {
            return DataAccess.Generator(typeName);
        }

        return Path.Combine(DataAccess.DataDir, typeName.ToLower());
    }

    protected byte[] OnLoadData(string typeName)
    {
        var datafile = OnGenerateFilename(typeName);

        if (localLoader != null)
        {
            return localLoader(datafile);
        }

        if (DataAccess.Loader != null)
        {
            return DataAccess.Loader(datafile);
        }

#if UNITY_ENGINE
        TextAsset text = Resources.Load<TextAsset>(dataPath);
		var bytes = text.bytes;
        GameObject.Destroy(text);
        return bytes;
#else
        var filename = datafile + DataAccess.DataExt;
        if (File.Exists(filename))
        {
            return File.ReadAllBytes(filename);
        }
        else
        {
            throw new Exception($"can not load data:{datafile}");
        }
#endif
    }

    public void Initial(LoadDataHandler loadHandle = null, FileNameGenerateHandler fileNameGenerateHandle = null)
    {
        localLoader = loadHandle;
        localGenerator = fileNameGenerateHandle;
    }

    static DataContainer<TItem> _instance;
    public static DataContainer<TItem> Instance
    {
        get
        {
            if (_instance == null)
            {
                _instance = new DataContainer<TItem>();
            }
            return _instance;
        }
    }

    public override void Clear()
    {
        _item = null;
    }

    public DataContainer<TItem> Preload(byte[] bytes = null)
    {
        if (_item == null)
        {
            _item = Load(bytes);
        }

        return this;
    }

    private TItem Load(byte[] bytes = null)
    {
        var type = typeof(TItem);
        bytes = bytes ?? OnLoadData(type.Name);

        using (var stream = new MemoryStream(bytes))
        {
            object data = Serializer.NonGeneric.Deserialize(type, stream);
            return data as TItem;
        }
    }

    TItem _item;
    public TItem Data
    {
        get
        {
            if (_item == null)
            {
                try
                {
                    _item = Load();
                }
                catch (Exception e)
                {
                    T_LOG.LogError(typeof(TItem).Name + e);
                    return null;
                }
            }
            return _item;
        }
        set
        {
            _item = value;
        }
    }
}

#if LUA_SUPPORT
public interface ILuaDataContainer<TID> where TID : IComparable
{
    int Count { get; }
    List<TID> IDs { get; }
    LuaTable GetTableByIndex(int index);
    LuaTable GetTable(TID ID);
}
#endif

public class DataContainer
{
    public virtual void Clear()
    {
        
    }
}

public partial class DataContainer<TID, TItem> : DataContainer<TItem>
#if LUA_SUPPORT
    , ILuaDataContainer<TID>
#endif
    where TID : IComparable
    where TItem : PBDataModel
{

    private List<TItem> Load(byte[] bytes = null)
    {
        var type = typeof(TItem);
        var arrTypeName = type.FullName + "_ARRAY";
        var arrayType = type.Assembly.GetType(arrTypeName);
        if (arrayType == null)
        {
            throw new Exception($"can not find data type:{arrTypeName}");
        }

        bytes = bytes ?? OnLoadData(type.Name);

        using (var stream = new MemoryStream(bytes))
        {
            object array = Serializer.NonGeneric.Deserialize(arrayType, stream);
            var pItems = arrayType.GetProperty("Items");
            return pItems.GetValue(array) as List<TItem>;
        }
    }

    static DataContainer<TID, TItem> _instance;
    public new static DataContainer<TID, TItem> Instance
    {
        get
        {
            if (_instance == null)
            {
                _instance = new DataContainer<TID, TItem>();
            }
            return _instance;
        }
    }

    public static DataContainer<TID, TItem> GetInstance(string keyName)
    {
        if (_instance == null)
        {
            _instance = new DataContainer<TID, TItem>(keyName);
        }
        return _instance;
    }

    string keyName;
    Dictionary<TID, TItem> _itemMap;
    public Dictionary<TID, TItem> ItemMap
    {
        get
        {
            if (_itemMap == null)
            {
                _itemMap = InitDataAsDict();
            }
            return _itemMap;
        }
        private set
        {
            _itemMap = value;
        }
    }
    
    public new DataContainer<TID, TItem> Preload(byte[] bytes = null)
    {
        if (_items == null)
        {
            _items = InitDataAsList(bytes);
        }

        return this;
    }

    List<TItem> _items;
    public List<TItem> Items
    {
        get
        {
            if (_items == null)
            {
                _items = InitDataAsList();
            }
            return _items;
        }
        private set
        {
            _items = value;
        }
    }

    public int Count
    {
        get
        {
            if (Items == null)
            {
                return 0;
            }

            return Items.Count;
        }
    }

    List<TID> _IDs;
    public List<TID> IDs
    {
        get
        {
            if (_IDs == null)
            {
                _IDs = new List<TID>(ItemMap.Keys);
            }
            return _IDs;
        }
        private set
        {
            _IDs = value;
        }
    }

    public TItem this[TID ID]
    {
        get
        {
            if (!ItemMap.ContainsKey(ID))
            {
                T_LOG.Log($"can not find id={ID} in {typeof(TItem).Name}");
                return default(TItem);
            }
            return ItemMap[ID];
        }
    }

    #if LUA_SUPPORT
    Dictionary<TID, LuaTable> luaTables = null;

    public LuaTable GetTableByIndex(int index)
    {
        if (index >= this.IDs.Count)
        {
            return null;
        }

        var id = this.IDs[index];
        return GetTable(id);
    }

    public LuaTable GetTable(TID ID)
    {
        if(DataAccess.CacheHashValue && luaTables == null)
        {
            luaTables = new Dictionary<TID, LuaTable>();
        }

        if (luaTables != null && luaTables.ContainsKey(ID))
        {
            return luaTables[ID];
        }
        else if(Contains(ID))
        {
            var ht = this[ID].ToLuaTable();
            if (luaTables != null)
            {
                luaTables[ID] = ht;
            }
            return ht;
        }
        return null;
    }
    #endif

    public DataContainer(string keyName = DataAccess.DefaultIDName)
    {
        this.keyName = keyName;
    }

    protected Dictionary<TID, TItem> InitDataAsDict()
    {
        Dictionary<TID, TItem> itemMap = null;
        try
        {
            var propID = typeof(TItem).GetProperty(keyName);
            if (propID == null)
            {
                throw new Exception($"can not find key:{keyName}");
            }
            itemMap = new Dictionary<TID, TItem>();
            var itr = Items.GetEnumerator();
            while (itr.MoveNext())
            {
                TID ID = (TID)propID.GetValue(itr.Current);
                if (ID == null || ID is string && ID as string == "")
                {
                    Debug.LogWarning("id can not be null or empty value!");
                    continue;
                }
                itemMap[ID] = itr.Current;
            }
            return itemMap;
        }
        catch (Exception e)
        {
            T_LOG.LogError(typeof(TItem).Name + e);
            return itemMap;
        }
    }

    protected List<TItem> InitDataAsList(byte[] bytes = null)
    {
        List<TItem> list = null;
        try
        {
            return Load(bytes);
        }
        catch (Exception e)
        {
            T_LOG.LogError(typeof(TItem).Name + e);
            return list;
        }
    }

    public bool Contains(TID ID)
    {
        if (ItemMap != null)
        {
            return ItemMap.ContainsKey(ID);
        }
        return false;
    }

    public override void Clear()
    {
        base.Clear();

        if (_itemMap != null)
        {
            _itemMap.Clear();
            _itemMap = null;
        }

        if (_items != null)
        {
            _items.Clear();
            _items = null;
        }
        
        #if LUA_SUPPORT
        if (luaTables != null)
        {
            foreach (var table in luaTables)
            {
                table.Value.Dispose();
            }

            luaTables = null;
        }
        #endif
    }
}

File: reader/typescript/DataAccess.ts
import { Asset, BufferAsset, resources } from "cc";
import * as $protobuf from "protobufjs";

export interface IMessage {
    create(props?:any);
    encode(message: any, writer?: $protobuf.Writer): $protobuf.Writer;
    encodeDelimited(message?: any, writer?: $protobuf.Writer): $protobuf.Writer;
    decode(reader: ($protobuf.Reader|Uint8Array), length?: number): any;
    decodeDelimited(reader: ($protobuf.Reader|Uint8Array)): any;
    verify(message: { [k: string]: any }): (string|null);
    fromObject(object: { [k: string]: any }): any;
    toObject(message: any, options?: $protobuf.IConversionOptions): { [k: string]: any };
    toJSON(): { [k: string]: any };
}

export interface IDataArray {
    Items: any[];
}

type FileNameGenerateHandler = (typeName:string)=>string;
type LoadDataHandler = (datafile: string) => Uint8Array;
type KeyType = number|string;
type DataKeyMap = {[key:KeyType]: any};

export class DataAccess
{
    /// <summary>
    /// 是否使用ProtoMember的tag作为hashtable的key
    /// </summary>
    public static useProtoMemberTagAsHashtableKey = false;
    public static cacheHashValue = true;
    public static dataExt = ".bin";

    /// <summary>
    /// 数据文件以assets为根目录的路径
    /// </summary>
    public static dataDir: string;
    public static generator: FileNameGenerateHandler;
    public static loader: LoadDataHandler;

    private static _items:{[key:string]:any} = {};
    private static _tables: {[key:string]:any} = {};

    public static initial(dataDir:string, loadHandle?:LoadDataHandler, fileNameGenerateHandle?:FileNameGenerateHandler){
        this.dataDir = dataDir;
        this.generator = fileNameGenerateHandle;
        this.loader = loadHandle;
    }

    /**
     * 获取配置表
     * @param dataType 配置的数据类型
     * @returns 
     */
    public static getDataItem<T>(dataType: new()=>T): DataItem<T> {
        let typename = dataType["__type_name__"];
        if(this._items[typename]) {
            return this._items[typename];
        }
        return this._items[typename] = new DataItem<T>(dataType);
    }

    /**
     * 获取配置表, 可自定义主键名称
     * @param dataType 配置的数据类型
     * @param keyName 主键名称
     * @returns 
     */
    public static getDataTable<T>(dataType:new()=>T, keyName = "ID") : DataTable<T> {
        let typename = dataType["__type_name__"];
        if(this._tables[typename]) {
            return this._tables[typename];
        }
        return this._tables[typename] = new DataTable<T>(dataType, keyName);
    }
}

export class DataItem<T> {
    protected localGenerator: FileNameGenerateHandler;
    protected localLoader: LoadDataHandler;

    protected dataType: new()=>T;
    protected source: BufferAsset;

    public constructor(dataType: new()=>T) {
        this.dataType = dataType;
    }

    private onGenerateFilename(typeName: string): string
    {
        if (this.localGenerator != null)
        {
            return this.localGenerator(typeName);
        }

        if (DataAccess.generator != null)
        {
            return DataAccess.generator(typeName);
        }

        typeName = typeName.replace("_ARRAY", "");
        return DataAccess.dataDir + typeName.toLocaleLowerCase();
    }

    protected onLoadData(typeName: string): Uint8Array
    {
        if (this.source != null)
        {
            return new Uint8Array(this.source.buffer());
        }

        var datafile = this.onGenerateFilename(typeName);

        if (this.localLoader != null)
        {
            return this.localLoader(datafile);
        }

        if (DataAccess.loader != null)
        {
            return DataAccess.loader(datafile);
        }

        this.source = resources.get<BufferAsset>(datafile, BufferAsset);
        if(this.source) {
            return new Uint8Array(this.source.buffer());
        }

        return null;
    }

    protected setSource(source: BufferAsset){
        this.source = source;
    }

    public initial(dataType:new()=>T, loadHandle?:LoadDataHandler, fileNameGenerateHandle?:FileNameGenerateHandler) : void {
        this.dataType = dataType;
        this.localLoader = loadHandle;
        this.localGenerator = fileNameGenerateHandle;
    }

    public static create<T>(dataType: new()=>T, source: BufferAsset): DataItem<T> {
        var instance = new DataItem<T>(dataType);
        instance.setSource(source);
        return instance;
    }

    public clear() {
        if (this.source != null)
        {
            this.source.destroy();
            this.source = null;
        }
    }

    protected load(): any {
        var buffer = this.onLoadData(this.dataType["__array_type_name__"]);
        var msgType:IMessage = this.dataType["__array_type__"] as any;   
        return msgType.decode(buffer);
    }

    private _item: T;
    public get data(): Readonly<T> {
        if (this._item == null)
        {
            this._item = this.load();
        }
        return this._item;
    }
}

export class DataTable<T> extends DataItem<T> {
    private _keyName = "ID";
    public get keyName() {
        return this._keyName;
    }

    constructor(dataType:new()=>T, keyName = "ID") {
        super(dataType);

        this._keyName = keyName;
    }

    protected load() : T[] {
        var arrTypeName = this.dataType["__array_type_name__"];

        var buffer = this.onLoadData(arrTypeName);
        var msgType:IMessage = this.dataType["getArrayType"]() as any;
        var message = msgType.decode(buffer);
        return (message as IDataArray).Items;        
    }

    public static create<T>(dataType: new()=>T, source: BufferAsset, keyName = "ID"): DataTable<T>{
        var instance = new DataTable<T>(dataType, keyName);
        instance.setSource(source);
        return instance;
    }

    private _itemMap: DataKeyMap;
    public get itemMap(): Readonly<DataKeyMap> {
        if (this._itemMap == null)
        {
            this._itemMap = this.initDataAsDict();
        }
        return this._itemMap;
    }

    private _items: T[];
    public get items(): Readonly<Readonly<T>[]> {
        try{
            if (this._items == null){
                this._items = this.initDataAsList();
            }
        }catch(e){
            console.log(`config data load error: ${e}`)
        }
        return this._items;
    }

    private _ids: KeyType[];
    public get IDs(): Readonly<KeyType[]>
    {
        if (this._ids == null)
        {
            this._ids = this.items.map((val, idx, arr)=>{
                //@ts-ignore
                arr.push(val[this.keyName]);
            }, []) as any as KeyType[];
        }
        return this._ids;
    }

    public getItem(key: KeyType): Readonly<T>
    {
        return this.itemMap[key];
    }

    protected initDataAsDict(): DataKeyMap {
        let itemMap: DataKeyMap = {};
        try{
            this.items.forEach((val)=>{
                itemMap[val[this.keyName]] = val;
            });
            return itemMap;
        }
        catch(e) {
            console.error(`can not get data map by key ${this.keyName}`);
            return itemMap;
        }
    }

    protected initDataAsList(): T[]{
        return this.load();
    }

    public contains(id: KeyType): boolean{
        return this.itemMap[id] != null;
    }

    public clear(){
        super.clear();

        this._itemMap = null;
        this._items = null;
    }
}
